### **ä¸€ã€æ•°æ®å‡†å¤‡**

#### 1. **ä¸‹è½½å¿…è¦æ–‡ä»¶**

|                       |                                                                     |                |                                  |
| --------------------- | ------------------------------------------------------------------- | -------------- | -------------------------------- |
| æ–‡ä»¶ç±»å‹                  |                                                                     | ç‰ˆæœ¬             | ä½œç”¨                               |
| **NRæ•°æ®åº“**             | **nr.gz**                                                           | **2024-02-07** | åŸå§‹è›‹ç™½åºåˆ—åº“                          |
| **åˆ†ç±»æ•°æ®åº“**             | **taxdump.tar.gz**                                                  | **2025-06-12** | å«nodes.dmpï¼ˆåˆ†ç±»å±‚çº§ï¼‰ã€names.dmpï¼ˆç‰©ç§åç§°ï¼‰ |
| **Accession-TaxIDæ˜ å°„** | **prot.accession2taxid.FULL.gz**<br><br>**prot.accession2taxid.gz** | **2025-06-06** | è›‹ç™½åºåˆ—ä¸TaxIDå¯¹åº”å…³ç³»<br><br>å®Œæ•´ç‰ˆå’Œç®€åŒ–ç‰ˆ    |

|                       |                                      |                |                                  |
| --------------------- | ------------------------------------ | -------------- | -------------------------------- |
| æ–‡ä»¶ç±»å‹                  |                                      | ç‰ˆæœ¬             | ä½œç”¨                               |
| **NRæ•°æ®åº“**             | **nr.gz**                            | **2024-02-07** | åŸå§‹è›‹ç™½åºåˆ—åº“                          |
| **åˆ†ç±»æ•°æ®åº“**             | **taxdump.tar.gz**                   | **2025-08-26** | å«nodes.dmpï¼ˆåˆ†ç±»å±‚çº§ï¼‰ã€names.dmpï¼ˆç‰©ç§åç§°ï¼‰ |
| **Accession-TaxIDæ˜ å°„** | **prot.accession2taxid.FULL.gz**<br> | **2025-08-22** | è›‹ç™½åºåˆ—ä¸TaxIDå¯¹åº”å…³ç³»                   |
|                       |                                      |                |                                  |
|                       |                                      |                |                                  |
|                       |                                      |                |                                  |
|                       |                                      |                |                                  |
|                       |                                      |                |                                  |
|                       |                                      |                |                                  |
microNR
```shell
# bacteria, archaea,virus, fungi
2,2157,10239,4751

mkdir microNR
taxonkit list --ids 2,2157,4751,10239 --indent "" > microNR.taxid

pigz -dc prot.accession2taxid.gz \
| csvtk grep -t -f taxid -P microNR.taxid \
| csvtk cut -t -f accession.version,taxid \
| sed 1d > microNR.acc2taxid


cat <(echo) <(pigz -dc nr.gz) \
| perl -e 'BEGIN{ $/ = "\n>"; <>; } while(<>){s/>$//; $i = index $_, "\n"; $h = substr $_, 0, $i; $s = substr $_, $i+1; if ($h !~ />/) { print ">$_"; next; }; $h = ">$h"; while($h =~ />([^ ]+ .+?) ?(?=>|$)/g){ $h1 = $1; $h1 =~ s/^\W+//; print ">$h1\n$s";} } ' \
| seqkit grep -f microNR.taxidt -o microNR.fa.gz

cut -f 1 $id.acc2taxid.txt > $id.acc.txt
```
https://scienceparkstudygroup.github.io/ibed-bioinformatics-page/source/core_tools/ncbi_nr.html
https://bioinf.shenwei.me/taxonkit/tutorial/

ğŸ’¡ **éªŒè¯å®Œæ•´æ€§**ï¼šä½¿ç”¨`md5sum -c *.md5`æ ¡éªŒæ–‡ä»¶ã€‚

```
https://ftp.ncbi.nih.gov/blast/db/FASTA/nr.gz
https://ftp.ncbi.nih.gov/blast/db/FASTA/nr.gz.md5

https://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz
https://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz.md5

https://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/prot.accession2taxid.FULL.gz
https://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/prot.accession2taxid.FULL.gz.md5

https://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/prot.accession2taxid.gz
https://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/prot.accession2taxid.gz.md5
```

diamond æ•°æ®åº“

```
mkdir taxdmp && tar -xf taxdump.tar.gz -C taxdump
diamond makedb \
--verbose \
--log \
--threads 47 \
--in nr.gz \
--db NR \
--taxonmap prot.accession2taxid.FULL.gz \
--taxonnodes taxdump/nodes.dmp \
--taxonnames taxdump/names.dmp
```

diamond + taxonkit

diamond å¯¹æ¯”

```
diamond blastp \
--db NR \
--taxonlist 2,2157,10239 \
--query protein.faa \
--threads 47 \
--log \
-â€“range-culling -k 25 -F 15 \
--tmpdir tmp \
--evalue 1e-03 \
--top 10 \
--sensitive \
--outfmt 6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore staxids sscinames \
--include-lineage \
--out diamond
```

1. `--range-culling`  
    è¿™æ˜¯ Diamond ä¸­ç”¨äºä¼˜åŒ–æ¯”å¯¹ç»“æœçš„ç­›é€‰å‚æ•°ï¼Œæ ¸å¿ƒä½œç”¨æ˜¯**å»é™¤åŒä¸€æŸ¥è¯¢åºåˆ—ä¸åŒä¸€ç›®æ ‡åºåˆ—ä¹‹é—´çš„é‡å åŒ¹é…**ã€‚å½“ä¸€ä¸ªæŸ¥è¯¢è›‹ç™½ä¸åŒä¸€ä¸ªç›®æ ‡è›‹ç™½å­˜åœ¨å¤šä¸ªå±€éƒ¨æ¯”å¯¹ç»“æœï¼ˆä¾‹å¦‚å› åºåˆ—åˆ†æ®µåŒ¹é…æˆ–é‡å¤ç»“æ„åŸŸå¯¼è‡´ï¼‰ï¼Œè¯¥å‚æ•°ä¼šä¿ç•™å¾—åˆ†æœ€é«˜çš„éé‡å åŒºåŸŸï¼Œå‰”é™¤å…¶ä»–é‡å çš„å†—ä½™åŒ¹é…ã€‚è¿™æœ‰åŠ©äºå‡å°‘ç»“æœä¸­çš„å†—ä½™ä¿¡æ¯ï¼Œç¡®ä¿æ¯ä¸ª â€œæŸ¥è¯¢ - ç›®æ ‡â€ å¯¹ä»…ä¿ç•™æœ€æœ‰ä»·å€¼çš„åŒ¹é…åŒºåŸŸï¼Œå°¤å…¶é€‚ç”¨äºåˆ†æå«é‡å¤ç»“æ„åŸŸçš„è›‹ç™½è´¨åºåˆ—ã€‚
    
2. `-k 25`  
    ç”¨äºæŒ‡å®š**æ¯ä¸ªæŸ¥è¯¢åºåˆ—ä¿ç•™çš„æœ€å¤§åŒ¹é…ç»“æœæ•°é‡**ï¼Œæ­¤å¤„è®¾ç½®ä¸º 25ï¼Œå³å¯¹è¾“å…¥æ–‡ä»¶ä¸­çš„æ¯ä¸ªè›‹ç™½åºåˆ—ï¼Œä»…ä¿ç•™å¾—åˆ†æœ€é«˜çš„å‰ 25 ä¸ªç›®æ ‡åºåˆ—åŒ¹é…ç»“æœã€‚è¯¥å‚æ•°å¹³è¡¡äº†ç»“æœçš„å…¨é¢æ€§ä¸è®¡ç®—æ•ˆç‡ï¼šå€¼è¿‡å¤§ä¼šå¯¼è‡´ç»“æœå†—ä½™ï¼ˆå°¤å…¶æ˜¯é«˜åº¦ä¿å®ˆçš„è›‹ç™½å¯èƒ½åŒ¹é…ä¸Šç™¾ä¸ªåŒæºåºåˆ—ï¼‰ï¼Œå€¼è¿‡å°å¯èƒ½é—æ¼é‡è¦åŒæºä¿¡æ¯ã€‚25 æ˜¯ä¸­é—´å€¼ï¼Œé€‚ç”¨äºå¤šæ•°éœ€å…¼é¡¾æ•æ„Ÿæ€§å’Œç®€æ´æ€§çš„åˆ†æåœºæ™¯ï¼ˆå¦‚åŠŸèƒ½æ³¨é‡Šåˆç­›ï¼‰ã€‚
    
3. `-F 15`  
    è¯¥å‚æ•°æ§åˆ¶åºåˆ—æ¯”å¯¹å‰çš„**ä½å¤æ‚åº¦åŒºåŸŸè¿‡æ»¤**ï¼Œæ•°å€¼ â€œ15â€ å¯¹åº” Diamond çš„å†…ç½®è¿‡æ»¤è§„åˆ™ç»„åˆï¼Œå…·ä½“ç­‰æ•ˆäºï¼š
    
    - å¯ç”¨å¯¹æŸ¥è¯¢åºåˆ—å’Œç›®æ ‡åºåˆ—ä¸­ä½å¤æ‚åº¦åŒºåŸŸï¼ˆå¦‚å¯Œå«æŸç§æ°¨åŸºé…¸çš„é‡å¤åºåˆ—ï¼‰çš„å±è”½ï¼ˆç±»ä¼¼ BLAST ä¸­çš„`-seg`æˆ–`-dust`å‚æ•°ï¼‰ï¼›
    - è¿™äº›åŒºåŸŸé€šå¸¸å› åºåˆ—ç»„æˆç‰¹æ®Šè€Œå®¹æ˜“äº§ç”Ÿå‡é˜³æ€§åŒ¹é…ï¼ˆéç”Ÿç‰©å­¦æ„ä¹‰çš„ç›¸ä¼¼æ€§ï¼‰ï¼Œè¿‡æ»¤åå¯æé«˜æ¯”å¯¹ç»“æœçš„å¯é æ€§ã€‚  
        è‹¥è®¾ç½®ä¸º`-F 0`åˆ™å…³é—­è¿‡æ»¤ï¼Œé€‚ç”¨äºéœ€è¦ä¿ç•™ä½å¤æ‚åº¦åŒºåŸŸï¼ˆå¦‚ç ”ç©¶ç‰¹å®šé‡å¤åºåˆ—åŠŸèƒ½ï¼‰çš„åœºæ™¯ï¼Œä½†å¯èƒ½å¢åŠ å‡é˜³æ€§ã€‚

```
taxonkit reformat \
--format "{d}\t{p}\t{c}\t{o}\t{f}\t{g}\t{s}" \
--data-dir taxdump \
--prefix-d "k__" \
--taxid-field 3 \
--fill-miss-rank \
--add-prefix \
--threads 16 \
--out-file taxonkit.tsv \
diamond.txt
```

åˆå¹¶ join

```
import dask.dataframe as dd
df_left = dd.read_csv("/home/data/wd/Vik/X207/Result/alig_result.xls",sep="\t",header=None)
df_right = dd.read_csv("/home/data/wd/Vik/X207/Result/prot.accession2taxid.FULL",sep="\t")
result = df_left.merge(df_right, left_on=1,right_on="accession.version", how="inner")
result.to_csv("/home/data/wd/Vik/X207/tmp/output_*.tsv", index=False, single_file=False,sep="\t",header=False)
cat output_*.tsv > acc_taxid.xls
awk -F'\t' '$3!=""' acc_taxid.xls > acc_taxid_filter.xls

zcat /home/data/wd/Vik/X207/DB/salmon.counts.xls.gz | head -n 3000 | \
awk 'BEGIN {FS=OFS="\t"} 
     NR==FNR {b[$1]=1; next} 
     FNR>1 && ($1 in b) {print $0}' /home/data/wd/Vik/X207/Result/taxnomy.xls - > /home/data/wd/Vik/X207/Result/count.xls

```

diamond + MEGAN

```
diamond blastp \
--db NR \
--taxonlist 2,2157,10239 \
--query protein.faa \
--threads 47 \
--log \
-â€“range-culling -k 25 -F 15 \
--tmpdir tmp \
--evalue 1e-03 \
--top 10 \
--outfmt 100 \
--sensitive \
--include-lineage \
--out diamond


diamond blastp \
--db /data/Alex/NR.dmnd \
--taxonlist 2,2157,10239 \
--query 03.X207.faa.rmdup.faa \
--threads 31 \
-b 0.5 \
--log \
--evalue 1e-05 \
--top 10 \
--outfmt 100 \
--sensitive \
--include-lineage \
--out diamond
```

```

wget https://software-ab.cs.uni-tuebingen.de/download/megan6/megan-map-Feb2022.db.zip

megan=/home/data/wd/Vik/APP/megan/tools
# diamond dda ->  megan dda
$megan/daa-meganizer \
--longReads \
--in /home/data/wd/Vik/X207/Result/X207.dda.daa \
--mapDB /home/data/wd/Vik/X207/DB/megan-map-Feb2022.db \
-lcaAlgorithm longReads \
--threads 55 --classify true

# megan dda -> megan rma
$megan/daa2rma \
--in /home/data/wd/Vik/X207/Result/X207.dda.daa \
--out /home/data/wd/Vik/X207/Result/X207.rma \
--mapDB /home/data/wd/Vik/X207/DB/megan-map-Feb2022.db \
--longReads --lcaAlgorithm longReads \
--threads 55 --classify true \
--minScore 50 --maxExpected 0.01 --topPercent 50 

# mapDB: MEGANæ˜ å°„åº“è·¯å¾„
# minScore: æœ€å°æ”¯æŒæ•°
# maxExpected: æœ€å¤§æœŸæœ›å€¼
# topPercent: ä¿ç•™Top Næ¯”å¯¹ç»“æœ
# longReads: å¯¹é•¿åºåˆ—ï¼ˆå¦‚contigsï¼‰å¯ç”¨ä¼˜åŒ–

# megan rma -> table
$megan/rma2info \
--in /home/data/wd/Vik/X207/Result/X207.rma \
--out /home/data/wd/Vik/X207/Result/X207.tsv \
--class2count Taxonomy --read2class Taxonomy \
--ignoreUnassigned true \
--names true --paths true --ranks true --list true --listMore true --majorRanksOnly true --verbose
# read2class: æŒ‡å®šè¾“å‡ºåˆ†ç±»ä¿¡æ¯
# names: æ˜¾ç¤ºç‰©ç§åç§°
# paths: è¾“å‡ºå®Œæ•´åˆ†ç±»è·¯å¾„ï¼ˆç•Œâ†’é—¨â†’å±â†’ç§ï¼‰
# ranks: åŒ…å«åˆ†ç±»å±‚çº§æ ‡è¯†
# list: ä»¥åˆ—è¡¨æ ¼å¼è¾“å‡º
# listMore: åŒ…å«æ›´å¤šä¿¡æ¯
# majorRanksOnly: ä»…ä¿ç•™ä¸»è¦åˆ†ç±»å±‚çº§ï¼ˆç•Œ/é—¨/çº²/ç›®/ç§‘/å±/ç§ï¼‰
# ignoreUnassigned: å¿½ç•¥æœªåˆ†ç±»çš„åºåˆ—
```

## Reference

[https://software-ab.cs.uni-tuebingen.de/download/megan6/welcome.html](https://software-ab.cs.uni-tuebingen.de/download/megan6/welcome.html)

[https://github.com/husonlab/megan-ce](https://github.com/husonlab/megan-ce)

[https://software-ab.cs.uni-tuebingen.de/download/megan6/manual.pdf](https://software-ab.cs.uni-tuebingen.de/download/megan6/manual.pdf)

[https://blog.csdn.net/woodcorpse/article/details/104381643](https://blog.csdn.net/woodcorpse/article/details/104381643)

https://www.jianshu.com/p/1d6edfcb4110